File: app/src/main/java/ru/yourok/flymefirmwarechecker/MainActivity.java
package ru.yourok.flymefirmwarechecker;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import org.json.JSONObject;

import ru.yourok.flymefirmwarechecker.dialogs.DateTimeDialog;
import ru.yourok.flymefirmwarechecker.dialogs.ItemsDialog;
import ru.yourok.utils.Params;
import ru.yourok.utils.RequestManager;
import ru.yourok.utils.SystemParams;
import ru.yourok.utils.Utils;

import static ru.yourok.utils.SystemParams.DEVICE_ID;
import static ru.yourok.utils.SystemParams.DEVICE_TYPE;
import static ru.yourok.utils.SystemParams.FIRMWARE;
import static ru.yourok.utils.SystemParams.IMEI;
import static ru.yourok.utils.SystemParams.ROOT;
import static ru.yourok.utils.SystemParams.SN;
import static ru.yourok.utils.SystemParams.SYSVER;
import static ru.yourok.utils.SystemParams.VERSION;

public class MainActivity extends AppCompatActivity {

    private JSONObject params;
    private JSONObject update;
    private RequestManager manager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ((CheckBox) findViewById(R.id.checkBoxInternational)).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setEditText(R.id.editTextHost, SystemParams.getHost(((CheckBox) findViewById(R.id.checkBoxInternational)).isChecked()));
            }
        });

        ((EditText) findViewById(R.id.editTextFirmType)).setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                final ItemsDialog itemsDialog = new ItemsDialog(MainActivity.this, R.array.firmware_type);
                itemsDialog.show(new Runnable() {
                    @Override
                    public void run() {
                        String item = itemsDialog.getItem();
                        if (item != null)
                            setEditText(R.id.editTextFirmType, item);
                    }
                });
                return true;
            }
        });

        View.OnLongClickListener cl = new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(final View v) {
                final ItemsDialog itemsDialog = new ItemsDialog(MainActivity.this, R.array.android_versions);
                itemsDialog.show(new Runnable() {
                    @Override
                    public void run() {
                        String item = itemsDialog.getItem();
                        if (item != null)
                            ((EditText) v).setText(item);
                    }
                });
                return true;
            }
        };

        ((EditText) findViewById(R.id.editTextVersion)).setOnLongClickListener(cl);
        ((EditText) findViewById(R.id.editTextFirmware)).setOnLongClickListener(cl);

        ((EditText) findViewById(R.id.editTextTimestamp)).setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                long timestamp = -1;
                if (!getEditText(R.id.editTextTimestamp).isEmpty())
                    timestamp = Long.parseLong(getEditText(R.id.editTextTimestamp));
                final DateTimeDialog dialog = new DateTimeDialog(MainActivity.this, timestamp);
                dialog.show().onConfirm(new Runnable() {
                    @Override
                    public void run() {
                        setEditText(R.id.editTextTimestamp, String.valueOf(dialog.getTimestamp()));
                    }
                });
                return true;
            }
        });

        ((EditText) findViewById(R.id.editTextDeviceType)).setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                final ItemsDialog itemsDialog = new ItemsDialog(MainActivity.this, R.array.devices_id);
                itemsDialog.show(new Runnable() {
                    @Override
                    public void run() {
                        String item = itemsDialog.getItem();
                        String[] tmp = item.split("-");
                        if (tmp.length == 2)
                            item = tmp[0];
                        setEditText(R.id.editTextDeviceType, item);
                    }
                });
                return true;
            }
        });
    }

    private void ShowMsg(final String msg) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Toast.makeText(MainActivity.this, msg, Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    protected void onStart() {
        try {
            Params p = new Params(this);
            params = p.getParams();
            if (!p.getUpdate().isEmpty())
                update = new JSONObject(p.getUpdate());
            if (params == null || (params.getString(SN).isEmpty() && params.getString(IMEI).isEmpty()))
                params = new SystemParams(this).getSystemParams();
            writeToUIParams(params);
        } catch (Exception e) {
            e.printStackTrace();
            ShowMsg(e.getMessage());
        }
        super.onStart();
    }

    @Override
    protected void onStop() {
        try {
            params = readFromUIParams();
            Params p = new Params(this);
            p.setParams(params);
            if (update != null)
                p.setUpdate(update.toString());
        } catch (Exception e) {
            e.printStackTrace();
            ShowMsg(e.getMessage());
        }
        super.onStop();
    }

    private String getEditText(int id) {
        if (id == R.id.checkBoxRoot) {
            if (((CheckBox) findViewById(id)).isChecked())
                return "1";
            return "0";
        }

        return ((EditText) findViewById(id)).getText().toString();
    }

    private void setEditText(final int id, final String val) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (id == R.id.checkBoxRoot)
                    ((CheckBox) findViewById(id)).setChecked(val.equals("1"));
                else
                    ((EditText) findViewById(id)).setText(val);
            }
        });
    }

    private JSONObject readFromUIParams() throws Exception {
        String andVer = getEditText(R.id.editTextVersion);
        String timestamp = getEditText(R.id.editTextTimestamp);
        String firmtype = getEditText(R.id.editTextFirmType);
        String maskid = andVer + "-" + timestamp + "_" + firmtype;

        JSONObject basicParams = new JSONObject();
        basicParams.put(DEVICE_TYPE, getEditText(R.id.editTextDeviceType));
        basicParams.put(FIRMWARE, getEditText(R.id.editTextFirmware));
        basicParams.put(ROOT, getEditText(R.id.checkBoxRoot));
        basicParams.put(SYSVER, maskid);
        basicParams.put(VERSION, maskid);
        basicParams.put(DEVICE_ID, getEditText(R.id.editTextImei));
        basicParams.put(SN, getEditText(R.id.editTextSN));
        basicParams.put(IMEI, getEditText(R.id.editTextImei));
        return basicParams;
    }

    private void writeToUIParams(JSONObject params) throws Exception {
        String maskid = params.getString(SYSVER);
        String[] prop = maskid.split("-|_");

        setEditText(R.id.editTextDeviceType, params.getString(DEVICE_TYPE));
        setEditText(R.id.editTextFirmware, params.getString(FIRMWARE));
        if (prop.length > 2) {
            setEditText(R.id.editTextVersion, prop[0]);
            setEditText(R.id.editTextTimestamp, prop[1]);
            setEditText(R.id.editTextFirmType, prop[2]);
        }
        setEditText(R.id.checkBoxRoot, params.getString(ROOT));
        setEditText(R.id.editTextImei, params.getString(DEVICE_ID));
        setEditText(R.id.editTextSN, params.getString(SN));
        UpdateView();
    }

    public void onBtnRead(View view) {
        try {
            params = new SystemParams(this).getSystemParams();
            writeToUIParams(params);
        } catch (Exception e) {
            e.printStackTrace();
            ShowMsg(e.getMessage());
        }
    }

    public void onBtnCheck(View view) {
        try {
            ((TextView) findViewById(R.id.textViewUpdate)).setText("Requesting...");
            params = readFromUIParams();

            final RequestManager manager = new RequestManager(getEditText(R.id.editTextHost), params);
            manager.check(new Runnable() {
                @Override
                public void run() {
                    final String stat = manager.getStat();
                    update = manager.getUpdate();
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            if (stat.isEmpty()) {
                                ShowMsg("Ok");
                                UpdateView();
                            } else
                                ((TextView) findViewById(R.id.textViewUpdate)).setText(stat);
                        }
                    });
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
            ShowMsg(e.getMessage());
        }
    }

    public void onBtnFind(View view) {
        try {
            if (manager != null && manager.isFinding()) {
                manager.stop();
                ((Button) findViewById(R.id.buttonFind)).setText("Find");
                return;
            }

            ((TextView) findViewById(R.id.textViewUpdate)).setText("Requesting...");
            params = readFromUIParams();

            final RequestManager rm = new RequestManager(getEditText(R.id.editTextHost), params);
            manager = rm;
            rm.find(new Runnable() {
                        @Override
                        public void run() {
                            final String stat = rm.getStat();
                            update = rm.getUpdate();
                            runOnUiThread(new Runnable() {
                                @Override
                                public void run() {
                                    if (stat == null)
                                        ((TextView) findViewById(R.id.textViewUpdate)).setText("");
                                    else {
                                        if (stat.isEmpty()) {
                                            ShowMsg("Ok");
                                            UpdateView();
                                        } else
                                            ((TextView) findViewById(R.id.textViewUpdate)).setText(stat);
                                    }
                                    ((Button) findViewById(R.id.buttonFind)).setText("Find");
                                }
                            });

                        }
                    },
                    new Runnable() {
                        @Override
                        public void run() {
                            long timestamp = rm.getTimestamp();
                            if (timestamp != -1) {
                                setEditText(R.id.editTextTimestamp, String.valueOf(timestamp));
                            }
                        }
                    });

            if (manager.isFinding())
                ((Button) findViewById(R.id.buttonFind)).setText("Stop");
            else
                ((Button) findViewById(R.id.buttonFind)).setText("Find");

        } catch (Exception e) {
            e.printStackTrace();
            ShowMsg(e.getMessage());
        }
    }

    private void UpdateView() {
        String upd = "";
        try {
            if (update != null) {
                if (Utils.has(update, "reply", "value", "new")) {
                    JSONObject jnew = update.getJSONObject("reply").getJSONObject("value").getJSONObject("new");
                    upd = jnew.getString("systemVersion") + " " + jnew.getString("fileSize") + " " + jnew.getString("releaseDate");
                } else if (update.has("code")) {
                    upd = update.getString("message");
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        ((TextView) findViewById(R.id.textViewUpdate)).setText(upd);
    }

    public void onBtnView(View view) {
        if (update == null) {
            ShowMsg("Nothing view");
            return;
        }
        Intent intent = new Intent(this, UpdateViewActivity.class);
        intent.putExtra("Update", update.toString());
        startActivity(intent);
    }

    public void onBtnEditor(View view) {
        Intent intent = new Intent(this, EditorActivity.class);
        startActivity(intent);
    }
}

File: app/src/main/java/ru/yourok/flymefirmwarechecker/MainActivity.kt
package ru.yourok.flymefirmwarechecker

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.*
import org.json.JSONObject
import ru.yourok.flymefirmwarechecker.dialogs.DateTimeDialog
import ru.yourok.flymefirmwarechecker.dialogs.ItemsDialog
import ru.yourok.utils.Params
import ru.yourok.utils.RequestManager
import ru.yourok.utils.SystemParams
import ru.yourok.utils.SystemParams.*
import ru.yourok.utils.Utils

class MainActivity : AppCompatActivity() {

    private var params: JSONObject? = null
    private var update: JSONObject? = null
    private var manager: RequestManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        (findViewById(R.id.checkBoxInternational) as CheckBox).setOnClickListener { setEditText(R.id.editTextHost, SystemParams.getHost((findViewById(R.id.checkBoxInternational) as CheckBox).isChecked)) }

        (findViewById(R.id.editTextFirmType) as EditText).setOnLongClickListener {
            val itemsDialog = ItemsDialog(this@MainActivity, R.array.firmware_type)
            itemsDialog.show {
                val item = itemsDialog.item
                if (item != null)
                    setEditText(R.id.editTextFirmType, item)
            }
            true
        }

        val cl = View.OnLongClickListener { v ->
            val itemsDialog = ItemsDialog(this@MainActivity, R.array.android_versions)
            itemsDialog.show {
                val item = itemsDialog.item
                if (item != null)
                    (v as EditText).setText(item)
            }
            true
        }

        (findViewById(R.id.editTextVersion) as EditText).setOnLongClickListener(cl)
        (findViewById(R.id.editTextFirmware) as EditText).setOnLongClickListener(cl)

        (findViewById(R.id.editTextTimestamp) as EditText).setOnLongClickListener {
            var timestamp: Long = -1
            if (!getEditText(R.id.editTextTimestamp).isEmpty())
                timestamp = java.lang.Long.parseLong(getEditText(R.id.editTextTimestamp))
            val dialog = DateTimeDialog(this@MainActivity, timestamp)
            dialog.show().onConfirm { setEditText(R.id.editTextTimestamp, dialog.timestamp.toString()) }
            true
        }

        (findViewById(R.id.editTextDeviceType) as EditText).setOnLongClickListener {
            val itemsDialog = ItemsDialog(this@MainActivity, R.array.devices_id)
            itemsDialog.show {
                var item = itemsDialog.item
                val tmp = item.split("-".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
                if (tmp.size == 2)
                    item = tmp[0]
                setEditText(R.id.editTextDeviceType, item)
            }
            true
        }
        (findViewById(R.id.textViewUpdate) as TextView).setOnClickListener {
            try {
                val jnew = update!!.getJSONObject("reply").getJSONObject("value").getJSONObject("new")
                val link = jnew.getString("updateUrl")
                if (!link.isNullOrEmpty()) {
                    val sendIntent = Intent(Intent.ACTION_VIEW, Uri.parse(link))
                    sendIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    val chooser = Intent.createChooser(sendIntent, "")
                    chooser.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    startActivity(chooser)
                }
            } catch (e: Exception) {

            }
        }
    }

    private fun ShowMsg(msg: String) {
        runOnUiThread { Toast.makeText(this@MainActivity, msg, Toast.LENGTH_SHORT).show() }
    }

    override fun onStart() {
        try {
            val p = Params(this)
            params = p.params
            if (!p.update.isEmpty())
                update = JSONObject(p.update)
            if (params == null || params!!.getString(SN).isEmpty() && params!!.getString(IMEI).isEmpty())
                params = SystemParams(this).systemParams
            writeToUIParams(params!!)
        } catch (e: Exception) {
            e.printStackTrace()
            e.message?.let {
                ShowMsg(it)
            }
        }

        super.onStart()
    }

    override fun onStop() {
        try {
            params = readFromUIParams()
            val p = Params(this)
            p.params = params
            if (update != null)
                p.update = update!!.toString()
        } catch (e: Exception) {
            e.printStackTrace()
            e.message?.let {
                ShowMsg(it)
            }
        }

        super.onStop()
    }

    private fun getEditText(id: Int): String {
        return if (id == R.id.checkBoxRoot) {
            if ((findViewById(id) as CheckBox).isChecked) "1" else "0"
        } else (findViewById(id) as EditText).text.toString()

    }

    private fun setEditText(id: Int, `val`: String?) {
        runOnUiThread {
            if (id == R.id.checkBoxRoot)
                (findViewById(id) as CheckBox).isChecked = `val` == "1"
            else
                (findViewById(id) as EditText).setText(`val`)
        }
    }

    private fun readFromUIParams(): JSONObject {
        val andVer = getEditText(R.id.editTextVersion)
        val timestamp = getEditText(R.id.editTextTimestamp)
        val firmtype = getEditText(R.id.editTextFirmType)
        val maskid = andVer + "-" + timestamp + "_" + firmtype

        val basicParams = JSONObject()
        basicParams.put(DEVICE_TYPE, getEditText(R.id.editTextDeviceType))
        basicParams.put(FIRMWARE, getEditText(R.id.editTextFirmware))
        basicParams.put(ROOT, getEditText(R.id.checkBoxRoot))
        basicParams.put(SYSVER, maskid)
        basicParams.put(VERSION, maskid)
        basicParams.put(DEVICE_ID, getEditText(R.id.editTextImei))
        basicParams.put(SN, getEditText(R.id.editTextSN))
        basicParams.put(IMEI, getEditText(R.id.editTextImei))
        return basicParams
    }

    private fun writeToUIParams(params: JSONObject) {
        val maskid = params.getString(SYSVER)
        val prop = maskid.split("-|_".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()

        setEditText(R.id.editTextDeviceType, params.getString(DEVICE_TYPE))
        setEditText(R.id.editTextFirmware, params.getString(FIRMWARE))
        if (prop.size > 2) {
            setEditText(R.id.editTextVersion, prop[0])
            setEditText(R.id.editTextTimestamp, prop[1])
            setEditText(R.id.editTextFirmType, prop[2])
        }
        setEditText(R.id.checkBoxRoot, params.getString(ROOT))
        setEditText(R.id.editTextImei, params.getString(DEVICE_ID))
        setEditText(R.id.editTextSN, params.getString(SN))
        UpdateView()
    }

    fun onBtnRead(view: View) {
        try {
            params = SystemParams(this).systemParams
            writeToUIParams(params!!)
        } catch (e: Exception) {
            e.printStackTrace()
            e.message?.let {
                ShowMsg(it)
            }
        }

    }

    fun onBtnCheck(view: View) {
        try {
            (findViewById(R.id.textViewUpdate) as TextView).text = "Requesting..."
            params = readFromUIParams()
            if (params == null) {
                ShowMsg("Set all params")
                return
            }
            val manager = RequestManager(getEditText(R.id.editTextHost), params!!)
            manager.check(Runnable {
                val stat = manager.stat
                update = manager.update
                runOnUiThread {
                    if (stat.isEmpty()) {
                        ShowMsg("Ok")
                        UpdateView()
                    } else
                        (findViewById(R.id.textViewUpdate) as TextView).text = stat
                }
            })
        } catch (e: Exception) {
            e.printStackTrace()
            e.message?.let {
                ShowMsg(it)
            }
        }

    }

    fun onBtnFind(view: View) {
        try {
            if (manager != null && manager!!.isFinding) {
                manager!!.stop()
                (findViewById(R.id.buttonFind) as Button).text = "Find"
                return
            }

            (findViewById(R.id.textViewUpdate) as TextView).text = "Requesting..."
            params = readFromUIParams()
            if (params == null) {
                ShowMsg("Set all params")
                return
            }
            val rm = RequestManager(getEditText(R.id.editTextHost), params!!)
            manager = rm
            rm.find(Runnable {
                val stat = rm.stat
                update = rm.update
                runOnUiThread {
                    if (stat.isEmpty()) {
                        ShowMsg("Ok")
                        UpdateView()
                    } else
                        (findViewById(R.id.textViewUpdate) as TextView).text = stat
                    (findViewById(R.id.buttonFind) as Button).text = "Find"
                }
            }, Runnable {
                val timestamp = rm.timestamp
                if (timestamp > -1) {
                    setEditText(R.id.editTextTimestamp, timestamp.toString())
                }
            })

            if (manager!!.isFinding)
                (findViewById(R.id.buttonFind) as Button).text = "Stop"
            else
                (findViewById(R.id.buttonFind) as Button).text = "Find"

        } catch (e: Exception) {
            e.printStackTrace()
            e.message?.let {
                ShowMsg(it)
            }
        }

    }

    private fun UpdateView() {
        var upd = ""
        try {
            if (update != null) {
                if (Utils.has(update, "reply", "value", "new")) {
                    val jnew = update!!.getJSONObject("reply").getJSONObject("value").getJSONObject("new")
                    upd = jnew.getString("systemVersion") + " " + jnew.getString("fileSize") + " " + jnew.getString("releaseDate")
                } else if (update!!.has("code")) {
                    upd = update!!.getString("message")
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        (findViewById(R.id.textViewUpdate) as TextView).text = upd
    }

    fun onBtnView(view: View) {
        if (update == null || !Utils.has(update, "reply", "value", "new")) {
            ShowMsg("Nothing view")
            return
        }
        val intent = Intent(this, UpdateViewActivity::class.java)
        intent.putExtra("Update", update!!.toString())
        startActivity(intent)
    }

    fun onBtnEditor(view: View) {
        val intent = Intent(this, EditorActivity::class.java)
        startActivity(intent)
    }
}
